### ISSUE TITLE
Bug in numeric conversions

### ISSUE THREAD

**Issue by GitMensch** [LADSoft/OrangeC#227]
@chuggafan pointed to tests for numeric conversions at http://www.tybor.com/.

I've actually run the [test programs](http://www.tybor.com/targzip.gz) and think the results should be fixed and afterwards the test programs tbin2dec.c, tsin.c and tint2flt.c should be added to the CI tests.

Note: running tflt2int.c (273 failures reported with current CI build) also fails when compiled with GCC 5.3.0 and 8.1.0 (linked with `-lm` which provide `fetestexcept()` and co) - 165 failures -. Therefore I won't add it to the tests and guess the number of failures may drop when the others are fixed.

Current results:

~~~
$ occ --nologo tbin2dec.c

$ tbin2dec.exe
Info:  'TAILOR' not defined => NOT using any bug workarounds.
Rounding is add 0.5 and truncate.
MIN_WIDTH=509.
Test ( Pass ) {}

$ occ --nologo tsin.c

$ tsin.exe
Info:  'TAILOR' not defined => NOT using any bug workarounds.
sin(355) = -3.01443533594889100e-05 from math library.
sin(355) = -3.01443533594884500e-05 is correct answer.
About 7.09 bits are wrong out of 53.00 total bits.
That is about 135.00 ULPs (Units Last Place) bits error.
Test Failed due to ULPs error being too large.

$ occ --nologo tflt2int.c

$ tflt2int.exe
Info:  'TAILOR' not defined => NOT using any bug workarounds.
This program tests conversions from floating-point types to integer types.
It tests positive powers of 2.0, e.g., f = 2.0**j
It tests both i = +f; and i = -f;
For numbers that are representable, it checks expected value and no exceptions.
For numbers that are NOT representable, it checks for FE_INVALID exception.
This implementation does not claim conformance to IEC-60559 or IEEE-754
Checks of the floating-point exception flags may be meaningless.
Testing conversion from long double to unsigned long long
-f: j=0, expected FE_INVALID, got: 0; i2 is 0X  ffffffffffffffff
-f: j=1, expected FE_INVALID, got: 0; i2 is 0X  fffffffffffffffe
-f: j=2, expected FE_INVALID, got: 0; i2 is 0X  fffffffffffffffc
-f: j=3, expected FE_INVALID, got: 0; i2 is 0X  fffffffffffffff8
-f: j=4, expected FE_INVALID, got: 0; i2 is 0X  fffffffffffffff0
-f: j=5, expected FE_INVALID, got: 0; i2 is 0X  ffffffffffffffe0
-f: j=6, expected FE_INVALID, got: 0; i2 is 0X  ffffffffffffffc0
-f: j=7, expected FE_INVALID, got: 0; i2 is 0X  ffffffffffffff80
-f: j=8, expected FE_INVALID, got: 0; i2 is 0X  ffffffffffffff00
Too many failures
Testing conversion from long double to signed long long
_ftoll: OVERFLOW error
+f: j=63, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
+f: j=64, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
-f: j=64, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
+f: j=65, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
-f: j=65, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
+f: j=66, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
-f: j=66, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
+f: j=67, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
-f: j=67, expected just FE_INVALID, got: 0X  41
Too many failures
Testing conversion from long double to unsigned long
-f: j=0, expected FE_INVALID, got: 0; i2 is 0X  ffffffff
-f: j=1, expected FE_INVALID, got: 0; i2 is 0X  fffffffe
-f: j=2, expected FE_INVALID, got: 0; i2 is 0X  fffffffc
-f: j=3, expected FE_INVALID, got: 0; i2 is 0X  fffffff8
-f: j=4, expected FE_INVALID, got: 0; i2 is 0X  fffffff0
-f: j=5, expected FE_INVALID, got: 0; i2 is 0X  ffffffe0
-f: j=6, expected FE_INVALID, got: 0; i2 is 0X  ffffffc0
-f: j=7, expected FE_INVALID, got: 0; i2 is 0X  ffffff80
-f: j=8, expected FE_INVALID, got: 0; i2 is 0X  ffffff00
Too many failures
Testing conversion from long double to signed long
+f: j=31, expected FE_INVALID, got: 0
+f: j=32, expected FE_INVALID, got: 0
-f: j=32, expected FE_INVALID, got: 0
+f: j=33, expected FE_INVALID, got: 0
-f: j=33, expected FE_INVALID, got: 0
+f: j=34, expected FE_INVALID, got: 0
-f: j=34, expected FE_INVALID, got: 0
+f: j=35, expected FE_INVALID, got: 0
-f: j=35, expected FE_INVALID, got: 0
Too many failures
Testing conversion from long double to unsigned int
-f: j=0, expected FE_INVALID, got: 0; i2 is 0X  ffffffff
-f: j=1, expected FE_INVALID, got: 0; i2 is 0X  fffffffe
-f: j=2, expected FE_INVALID, got: 0; i2 is 0X  fffffffc
-f: j=3, expected FE_INVALID, got: 0; i2 is 0X  fffffff8
-f: j=4, expected FE_INVALID, got: 0; i2 is 0X  fffffff0
-f: j=5, expected FE_INVALID, got: 0; i2 is 0X  ffffffe0
-f: j=6, expected FE_INVALID, got: 0; i2 is 0X  ffffffc0
-f: j=7, expected FE_INVALID, got: 0; i2 is 0X  ffffff80
-f: j=8, expected FE_INVALID, got: 0; i2 is 0X  ffffff00
Too many failures
Testing conversion from long double to signed int
+f: j=31, expected FE_INVALID, got: 0
+f: j=32, expected FE_INVALID, got: 0
-f: j=32, expected FE_INVALID, got: 0
+f: j=33, expected FE_INVALID, got: 0
-f: j=33, expected FE_INVALID, got: 0
+f: j=34, expected FE_INVALID, got: 0
-f: j=34, expected FE_INVALID, got: 0
+f: j=35, expected FE_INVALID, got: 0
-f: j=35, expected FE_INVALID, got: 0
Too many failures
Testing conversion from long double to unsigned short
-f: j=0, expected FE_INVALID, got: 0; i2 is 0X  ffff
-f: j=1, expected FE_INVALID, got: 0; i2 is 0X  fffe
-f: j=2, expected FE_INVALID, got: 0; i2 is 0X  fffc
-f: j=3, expected FE_INVALID, got: 0; i2 is 0X  fff8
-f: j=4, expected FE_INVALID, got: 0; i2 is 0X  fff0
-f: j=5, expected FE_INVALID, got: 0; i2 is 0X  ffe0
-f: j=6, expected FE_INVALID, got: 0; i2 is 0X  ffc0
-f: j=7, expected FE_INVALID, got: 0; i2 is 0X  ff80
-f: j=8, expected FE_INVALID, got: 0; i2 is 0X  ff00
Too many failures
Testing conversion from long double to signed short
+f: j=15, expected FE_INVALID, got: 0
+f: j=16, expected FE_INVALID, got: 0
-f: j=16, expected FE_INVALID, got: 0
+f: j=17, expected FE_INVALID, got: 0
-f: j=17, expected FE_INVALID, got: 0
+f: j=18, expected FE_INVALID, got: 0
-f: j=18, expected FE_INVALID, got: 0
+f: j=19, expected FE_INVALID, got: 0
-f: j=19, expected FE_INVALID, got: 0
Too many failures
Testing conversion from long double to unsigned char
-f: j=0, expected FE_INVALID, got: 0; i2 is 0X  ff
-f: j=1, expected FE_INVALID, got: 0; i2 is 0X  fe
-f: j=2, expected FE_INVALID, got: 0; i2 is 0X  fc
-f: j=3, expected FE_INVALID, got: 0; i2 is 0X  f8
-f: j=4, expected FE_INVALID, got: 0; i2 is 0X  f0
-f: j=5, expected FE_INVALID, got: 0; i2 is 0X  e0
-f: j=6, expected FE_INVALID, got: 0; i2 is 0X  c0
-f: j=7, expected FE_INVALID, got: 0; i2 is 0X  80
+f: j=8, expected FE_INVALID, got: 0
-f: j=8, expected FE_INVALID, got: 0; i2 is 0
Too many failures
Testing conversion from long double to signed char
+f: j=7, expected FE_INVALID, got: 0
+f: j=8, expected FE_INVALID, got: 0
-f: j=8, expected FE_INVALID, got: 0
+f: j=9, expected FE_INVALID, got: 0
-f: j=9, expected FE_INVALID, got: 0
+f: j=10, expected FE_INVALID, got: 0
-f: j=10, expected FE_INVALID, got: 0
+f: j=11, expected FE_INVALID, got: 0
-f: j=11, expected FE_INVALID, got: 0
Too many failures
Testing conversion from double to unsigned long long
-f: j=0, expected FE_INVALID, got: 0; i2 is 0X  ffffffffffffffff
-f: j=1, expected FE_INVALID, got: 0; i2 is 0X  fffffffffffffffe
-f: j=2, expected FE_INVALID, got: 0; i2 is 0X  fffffffffffffffc
-f: j=3, expected FE_INVALID, got: 0; i2 is 0X  fffffffffffffff8
-f: j=4, expected FE_INVALID, got: 0; i2 is 0X  fffffffffffffff0
-f: j=5, expected FE_INVALID, got: 0; i2 is 0X  ffffffffffffffe0
-f: j=6, expected FE_INVALID, got: 0; i2 is 0X  ffffffffffffffc0
-f: j=7, expected FE_INVALID, got: 0; i2 is 0X  ffffffffffffff80
-f: j=8, expected FE_INVALID, got: 0; i2 is 0X  ffffffffffffff00
Too many failures
Testing conversion from double to signed long long
_ftoll: OVERFLOW error
+f: j=63, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
+f: j=64, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
-f: j=64, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
+f: j=65, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
-f: j=65, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
+f: j=66, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
-f: j=66, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
+f: j=67, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
-f: j=67, expected just FE_INVALID, got: 0X  41
Too many failures
Testing conversion from double to unsigned long
-f: j=0, expected FE_INVALID, got: 0; i2 is 0X  ffffffff
-f: j=1, expected FE_INVALID, got: 0; i2 is 0X  fffffffe
-f: j=2, expected FE_INVALID, got: 0; i2 is 0X  fffffffc
-f: j=3, expected FE_INVALID, got: 0; i2 is 0X  fffffff8
-f: j=4, expected FE_INVALID, got: 0; i2 is 0X  fffffff0
-f: j=5, expected FE_INVALID, got: 0; i2 is 0X  ffffffe0
-f: j=6, expected FE_INVALID, got: 0; i2 is 0X  ffffffc0
-f: j=7, expected FE_INVALID, got: 0; i2 is 0X  ffffff80
-f: j=8, expected FE_INVALID, got: 0; i2 is 0X  ffffff00
Too many failures
Testing conversion from double to signed long
+f: j=31, expected FE_INVALID, got: 0
+f: j=32, expected FE_INVALID, got: 0
-f: j=32, expected FE_INVALID, got: 0
+f: j=33, expected FE_INVALID, got: 0
-f: j=33, expected FE_INVALID, got: 0
+f: j=34, expected FE_INVALID, got: 0
-f: j=34, expected FE_INVALID, got: 0
+f: j=35, expected FE_INVALID, got: 0
-f: j=35, expected FE_INVALID, got: 0
Too many failures
Testing conversion from double to unsigned int
-f: j=0, expected FE_INVALID, got: 0; i2 is 0X  ffffffff
-f: j=1, expected FE_INVALID, got: 0; i2 is 0X  fffffffe
-f: j=2, expected FE_INVALID, got: 0; i2 is 0X  fffffffc
-f: j=3, expected FE_INVALID, got: 0; i2 is 0X  fffffff8
-f: j=4, expected FE_INVALID, got: 0; i2 is 0X  fffffff0
-f: j=5, expected FE_INVALID, got: 0; i2 is 0X  ffffffe0
-f: j=6, expected FE_INVALID, got: 0; i2 is 0X  ffffffc0
-f: j=7, expected FE_INVALID, got: 0; i2 is 0X  ffffff80
-f: j=8, expected FE_INVALID, got: 0; i2 is 0X  ffffff00
Too many failures
Testing conversion from double to signed int
+f: j=31, expected FE_INVALID, got: 0
+f: j=32, expected FE_INVALID, got: 0
-f: j=32, expected FE_INVALID, got: 0
+f: j=33, expected FE_INVALID, got: 0
-f: j=33, expected FE_INVALID, got: 0
+f: j=34, expected FE_INVALID, got: 0
-f: j=34, expected FE_INVALID, got: 0
+f: j=35, expected FE_INVALID, got: 0
-f: j=35, expected FE_INVALID, got: 0
Too many failures
Testing conversion from double to unsigned short
-f: j=0, expected FE_INVALID, got: 0; i2 is 0X  ffff
-f: j=1, expected FE_INVALID, got: 0; i2 is 0X  fffe
-f: j=2, expected FE_INVALID, got: 0; i2 is 0X  fffc
-f: j=3, expected FE_INVALID, got: 0; i2 is 0X  fff8
-f: j=4, expected FE_INVALID, got: 0; i2 is 0X  fff0
-f: j=5, expected FE_INVALID, got: 0; i2 is 0X  ffe0
-f: j=6, expected FE_INVALID, got: 0; i2 is 0X  ffc0
-f: j=7, expected FE_INVALID, got: 0; i2 is 0X  ff80
-f: j=8, expected FE_INVALID, got: 0; i2 is 0X  ff00
Too many failures
Testing conversion from double to signed short
+f: j=15, expected FE_INVALID, got: 0
+f: j=16, expected FE_INVALID, got: 0
-f: j=16, expected FE_INVALID, got: 0
+f: j=17, expected FE_INVALID, got: 0
-f: j=17, expected FE_INVALID, got: 0
+f: j=18, expected FE_INVALID, got: 0
-f: j=18, expected FE_INVALID, got: 0
+f: j=19, expected FE_INVALID, got: 0
-f: j=19, expected FE_INVALID, got: 0
Too many failures
Testing conversion from double to unsigned char
-f: j=0, expected FE_INVALID, got: 0; i2 is 0X  ff
-f: j=1, expected FE_INVALID, got: 0; i2 is 0X  fe
-f: j=2, expected FE_INVALID, got: 0; i2 is 0X  fc
-f: j=3, expected FE_INVALID, got: 0; i2 is 0X  f8
-f: j=4, expected FE_INVALID, got: 0; i2 is 0X  f0
-f: j=5, expected FE_INVALID, got: 0; i2 is 0X  e0
-f: j=6, expected FE_INVALID, got: 0; i2 is 0X  c0
-f: j=7, expected FE_INVALID, got: 0; i2 is 0X  80
+f: j=8, expected FE_INVALID, got: 0
-f: j=8, expected FE_INVALID, got: 0; i2 is 0
Too many failures
Testing conversion from double to signed char
+f: j=7, expected FE_INVALID, got: 0
+f: j=8, expected FE_INVALID, got: 0
-f: j=8, expected FE_INVALID, got: 0
+f: j=9, expected FE_INVALID, got: 0
-f: j=9, expected FE_INVALID, got: 0
+f: j=10, expected FE_INVALID, got: 0
-f: j=10, expected FE_INVALID, got: 0
+f: j=11, expected FE_INVALID, got: 0
-f: j=11, expected FE_INVALID, got: 0
Too many failures
Testing conversion from float to unsigned long long
-f: j=0, expected FE_INVALID, got: 0; i2 is 0X  ffffffffffffffff
-f: j=1, expected FE_INVALID, got: 0; i2 is 0X  fffffffffffffffe
-f: j=2, expected FE_INVALID, got: 0; i2 is 0X  fffffffffffffffc
-f: j=3, expected FE_INVALID, got: 0; i2 is 0X  fffffffffffffff8
-f: j=4, expected FE_INVALID, got: 0; i2 is 0X  fffffffffffffff0
-f: j=5, expected FE_INVALID, got: 0; i2 is 0X  ffffffffffffffe0
-f: j=6, expected FE_INVALID, got: 0; i2 is 0X  ffffffffffffffc0
-f: j=7, expected FE_INVALID, got: 0; i2 is 0X  ffffffffffffff80
-f: j=8, expected FE_INVALID, got: 0; i2 is 0X  ffffffffffffff00
Too many failures
Testing conversion from float to signed long long
_ftoll: OVERFLOW error
+f: j=63, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
+f: j=64, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
-f: j=64, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
+f: j=65, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
-f: j=65, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
+f: j=66, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
-f: j=66, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
+f: j=67, expected just FE_INVALID, got: 0X  41
_ftoll: OVERFLOW error
-f: j=67, expected just FE_INVALID, got: 0X  41
Too many failures
Testing conversion from float to unsigned long
-f: j=0, expected FE_INVALID, got: 0; i2 is 0X  ffffffff
-f: j=1, expected FE_INVALID, got: 0; i2 is 0X  fffffffe
-f: j=2, expected FE_INVALID, got: 0; i2 is 0X  fffffffc
-f: j=3, expected FE_INVALID, got: 0; i2 is 0X  fffffff8
-f: j=4, expected FE_INVALID, got: 0; i2 is 0X  fffffff0
-f: j=5, expected FE_INVALID, got: 0; i2 is 0X  ffffffe0
-f: j=6, expected FE_INVALID, got: 0; i2 is 0X  ffffffc0
-f: j=7, expected FE_INVALID, got: 0; i2 is 0X  ffffff80
-f: j=8, expected FE_INVALID, got: 0; i2 is 0X  ffffff00
Too many failures
Testing conversion from float to signed long
+f: j=31, expected FE_INVALID, got: 0
+f: j=32, expected FE_INVALID, got: 0
-f: j=32, expected FE_INVALID, got: 0
+f: j=33, expected FE_INVALID, got: 0
-f: j=33, expected FE_INVALID, got: 0
+f: j=34, expected FE_INVALID, got: 0
-f: j=34, expected FE_INVALID, got: 0
+f: j=35, expected FE_INVALID, got: 0
-f: j=35, expected FE_INVALID, got: 0
Too many failures
Testing conversion from float to unsigned int
-f: j=0, expected FE_INVALID, got: 0; i2 is 0X  ffffffff
-f: j=1, expected FE_INVALID, got: 0; i2 is 0X  fffffffe
-f: j=2, expected FE_INVALID, got: 0; i2 is 0X  fffffffc
-f: j=3, expected FE_INVALID, got: 0; i2 is 0X  fffffff8
-f: j=4, expected FE_INVALID, got: 0; i2 is 0X  fffffff0
-f: j=5, expected FE_INVALID, got: 0; i2 is 0X  ffffffe0
-f: j=6, expected FE_INVALID, got: 0; i2 is 0X  ffffffc0
-f: j=7, expected FE_INVALID, got: 0; i2 is 0X  ffffff80
-f: j=8, expected FE_INVALID, got: 0; i2 is 0X  ffffff00
Too many failures
Testing conversion from float to signed int
+f: j=31, expected FE_INVALID, got: 0
+f: j=32, expected FE_INVALID, got: 0
-f: j=32, expected FE_INVALID, got: 0
+f: j=33, expected FE_INVALID, got: 0
-f: j=33, expected FE_INVALID, got: 0
+f: j=34, expected FE_INVALID, got: 0
-f: j=34, expected FE_INVALID, got: 0
+f: j=35, expected FE_INVALID, got: 0
-f: j=35, expected FE_INVALID, got: 0
Too many failures
Testing conversion from float to unsigned short
-f: j=0, expected FE_INVALID, got: 0; i2 is 0X  ffff
-f: j=1, expected FE_INVALID, got: 0; i2 is 0X  fffe
-f: j=2, expected FE_INVALID, got: 0; i2 is 0X  fffc
-f: j=3, expected FE_INVALID, got: 0; i2 is 0X  fff8
-f: j=4, expected FE_INVALID, got: 0; i2 is 0X  fff0
-f: j=5, expected FE_INVALID, got: 0; i2 is 0X  ffe0
-f: j=6, expected FE_INVALID, got: 0; i2 is 0X  ffc0
-f: j=7, expected FE_INVALID, got: 0; i2 is 0X  ff80
-f: j=8, expected FE_INVALID, got: 0; i2 is 0X  ff00
Too many failures
Testing conversion from float to signed short
+f: j=15, expected FE_INVALID, got: 0
+f: j=16, expected FE_INVALID, got: 0
-f: j=16, expected FE_INVALID, got: 0
+f: j=17, expected FE_INVALID, got: 0
-f: j=17, expected FE_INVALID, got: 0
+f: j=18, expected FE_INVALID, got: 0
-f: j=18, expected FE_INVALID, got: 0
+f: j=19, expected FE_INVALID, got: 0
-f: j=19, expected FE_INVALID, got: 0
Too many failures
Testing conversion from float to unsigned char
-f: j=0, expected FE_INVALID, got: 0; i2 is 0X  ff
-f: j=1, expected FE_INVALID, got: 0; i2 is 0X  fe
-f: j=2, expected FE_INVALID, got: 0; i2 is 0X  fc
-f: j=3, expected FE_INVALID, got: 0; i2 is 0X  f8
-f: j=4, expected FE_INVALID, got: 0; i2 is 0X  f0
-f: j=5, expected FE_INVALID, got: 0; i2 is 0X  e0
-f: j=6, expected FE_INVALID, got: 0; i2 is 0X  c0
-f: j=7, expected FE_INVALID, got: 0; i2 is 0X  80
+f: j=8, expected FE_INVALID, got: 0
-f: j=8, expected FE_INVALID, got: 0; i2 is 0
Too many failures
Testing conversion from float to signed char
+f: j=7, expected FE_INVALID, got: 0
+f: j=8, expected FE_INVALID, got: 0
-f: j=8, expected FE_INVALID, got: 0
+f: j=9, expected FE_INVALID, got: 0
-f: j=9, expected FE_INVALID, got: 0
+f: j=10, expected FE_INVALID, got: 0
-f: j=10, expected FE_INVALID, got: 0
+f: j=11, expected FE_INVALID, got: 0
-f: j=11, expected FE_INVALID, got: 0
Too many failures
To help you figure out which flags were raised:
 FE_INVALID   = 0X  1
 FE_DIVBYZERO = 0X  4
 FE_OVERFLOW  = 0X  8
 FE_UNDERFLOW = 0X  10
 FE_INEXACT   = 0X  20

... Symbols defined ...
Test Failed with at least 273 failures

$ occ --nologo tint2flt.c

$ tint2flt.exe
Info:  'TAILOR' not defined => NOT using any bug workarounds.
This program tests conversions from integer types to floating-point types.
For all types, it test positive powers of 2, e.g., 2**j
For signed types, it also tests negative of positive powers of 2, e.g., -(2**j)
For numbers that are representable, it checks expected value and no exceptions.
This implementation does not claim conformance to IEC-60559 or IEEE-754
Checks of the floating-point exception flags may be meaningless.
Testing conversion from unsigned long long to long double
j=31, expected power of 2, got: -2147483648.
j=32, expected power of 2, got: 0.
j=33, expected power of 2, got: 0.
j=34, expected power of 2, got: 0.
j=35, expected power of 2, got: 0.
j=36, expected power of 2, got: 0.
j=37, expected power of 2, got: 0.
j=38, expected power of 2, got: 0.
j=39, expected power of 2, got: 0.
Too many failures
Testing conversion from signed long long to long double
j=31, expected power of 2, got: -2147483648.
j=32, expected power of 2, got: 0.
j=32, expected negative of power of 2, got: 0.
j=33, expected power of 2, got: 0.
j=33, expected negative of power of 2, got: 0.
j=34, expected power of 2, got: 0.
j=34, expected negative of power of 2, got: 0.
j=35, expected power of 2, got: 0.
j=35, expected negative of power of 2, got: 0.
Too many failures
Testing conversion from unsigned long to long double
j=31, expected power of 2, got: -2147483648.
Testing conversion from signed long to long double
Testing conversion from unsigned int to long double
j=31, expected power of 2, got: -2147483648.
Testing conversion from signed int to long double
Testing conversion from unsigned short to long double
Testing conversion from signed short to long double
Testing conversion from unsigned char to long double
Testing conversion from signed char to long double
Testing conversion from unsigned long long to double
j=31, expected power of 2, got: -2147483648.
j=32, expected power of 2, got: 0.
j=33, expected power of 2, got: 0.
j=34, expected power of 2, got: 0.
j=35, expected power of 2, got: 0.
j=36, expected power of 2, got: 0.
j=37, expected power of 2, got: 0.
j=38, expected power of 2, got: 0.
j=39, expected power of 2, got: 0.
Too many failures
Testing conversion from signed long long to double
j=31, expected power of 2, got: -2147483648.
j=32, expected power of 2, got: 0.
j=32, expected negative of power of 2, got: 0.
j=33, expected power of 2, got: 0.
j=33, expected negative of power of 2, got: 0.
j=34, expected power of 2, got: 0.
j=34, expected negative of power of 2, got: 0.
j=35, expected power of 2, got: 0.
j=35, expected negative of power of 2, got: 0.
Too many failures
Testing conversion from unsigned long to double
j=31, expected power of 2, got: -2147483648.
Testing conversion from signed long to double
Testing conversion from unsigned int to double
j=31, expected power of 2, got: -2147483648.
Testing conversion from signed int to double
Testing conversion from unsigned short to double
Testing conversion from signed short to double
Testing conversion from unsigned char to double
Testing conversion from signed char to double
Testing conversion from unsigned long long to float
j=31, expected power of 2, got: -2147483648.
j=32, expected power of 2, got: 0.
j=33, expected power of 2, got: 0.
j=34, expected power of 2, got: 0.
j=35, expected power of 2, got: 0.
j=36, expected power of 2, got: 0.
j=37, expected power of 2, got: 0.
j=38, expected power of 2, got: 0.
j=39, expected power of 2, got: 0.
Too many failures
Testing conversion from signed long long to float
j=31, expected power of 2, got: -2147483648.
j=32, expected power of 2, got: 0.
j=32, expected negative of power of 2, got: 0.
j=33, expected power of 2, got: 0.
j=33, expected negative of power of 2, got: 0.
j=34, expected power of 2, got: 0.
j=34, expected negative of power of 2, got: 0.
j=35, expected power of 2, got: 0.
j=35, expected negative of power of 2, got: 0.
Too many failures
Testing conversion from unsigned long to float
j=31, expected power of 2, got: -2147483648.
Testing conversion from signed long to float
Testing conversion from unsigned int to float
j=31, expected power of 2, got: -2147483648.
Testing conversion from signed int to float
Testing conversion from unsigned short to float
Testing conversion from signed short to float
Testing conversion from unsigned char to float
Testing conversion from signed char to float
To help you figure out which flags were raised:
 FE_INVALID   = 0X  1
 FE_DIVBYZERO = 0X  4
 FE_OVERFLOW  = 0X  8
 FE_UNDERFLOW = 0X  10
 FE_INEXACT   = 0X  20

... Symbols defined ...
Test Failed with at least 60 failures
~~~

I haven't checked how to setup the FPCE Test Suite (with the sample configuration occ complains about undefined symbols ldbl_pair, ldc_got, ldc_want, dbl_pair, dc_got, dc_want, flt_pair, fc_got, fc_want during linking, gcc complains about unkown type names ldbl_pair, dbl_pair, flt_pair).

For the CI build I see two options to let it fail if the tests fail:

1) add the unchanged sources, in CI: compile, run the result and ogrep for "Fail", do something with the result
2) change the sources to return 1 if a tests failed, in CI: compile, run and let omake because the return is nonzero

**Comment by LADSoft** [comment]
sorry i'm getting behind, #222 turned into a big deal with a lot of compiler bugs...   I have one more to fix then will get back to other stuff for a while

**Comment by chuggafan** [comment]
If that requires an assembler rewrite to support SSE it may as well be a re-write to support x64 aswell (for the assembler, the compiler might take quite a bit longer to do that), http://ref.x86asm.net/geek64.html provides a nice (nearly complete) list of assembler opcodes, but if you want a (full, hard to read and convert to the adl format you're using) list: https://github.com/intelxed/xed intel's got you covered for both AMD and intel.

**Comment by LADSoft** [comment]
thanks for the links.   I reviewed and the x64 support seems minimal compared to the SSE support, so, makes sense to do both while I'm in there.

This is fairly high on my list but I've got some other stuff in front of it.

**Comment by chuggafan** [comment]
Any word on how you're doing for that assembler re-write? I might look into using NASM's x86.dat file and parsing out of there into the format you have if needed.

**Comment by LADSoft** [comment]
i have gotten nowhere on that...   right now I'm trying to dig out.   The current problem is that I can't get travis-ci to trigger builds...   I had disabled it so I wouldn't get millions of notifications but can't get it back.

**Comment by GitMensch** [comment]
As this is somehow the Travis issue now, too - I'll throw in https://github.com/marketplace/azure-pipelines/ - just out of curiosity...
@LADSoft  What do you think about registering and use the integration to build OrangeC (providing also minimal checks)? Azure has very likely much more performance than Appveyor will ever provide...

**Comment by LADSoft** [comment]
I guess I dropped the ball, both the Travis/CI and the CompileViaAssembly ideas should have had their own issues.   CompileViaAssembly seems to work on appveyor as of tonite; I may yet open an issue for Travis/CI depending on how much work there is left to do.

**Comment by GitMensch** [comment]
I actually would like to see the numeric conversions fixed ;-)

**Comment by LADSoft** [comment]
i'm starting to work on this now :)

**Comment by LADSoft** [comment]
over the break got oasm to properly generate almost all possible byte sequences for variations on the 'not' instruction in the three processor modes...   that is something like 110000 byte sequences....    have written extensive tests for the integer and floating point instructions (which assemble in NASM) but haven't tried OASM with them yet.   Going to finish the test scripts for SSE instructions then start testing in earnest...

**Comment by LADSoft** [comment]
got oasm rewritten and fully validated it over the last weekend...   but as of right now the ci build would fail because of a bug the rewrite found in dlpe.   Sigh.

**Comment by GitMensch** [comment]
> got oasm rewritten and fully validated it over the last weekend...

Yay? Do the test programs bin2dec.c, tsin.c and tint2flt.c referenced in the first post result in a reasonable error-free output? Are those programs compiled and running with checked results in the CI build already?

**Comment by LADSoft** [comment]
I wish it were that easy lol!  Still to come, a compiler rewrite to output SSE instructions, and a library rewrite to use SSE for floating point calculations :).    This is a fairly large project even once the assembler is working!

**Comment by chuggafan** [comment]
Sounds entirely good to me, if you're looking for code that you can modify for use in the RTL, glibc has you perfectly, musl, newlib, and bsd libc all are pretty good in terms of licencing, with the following licenses:

Glibc: https://sourceware.org/git/?p=glibc.git;a=blob;f=LICENSES;h=0e3a9fe39b26e97038d92f904508a4c3aa1bb43b;hb=HEAD with a majority splash of the GPL 2.1 or later.

musl: https://git.musl-libc.org/cgit/musl/tree/COPYRIGHT MIT license, supports re-licensing

newlib: https://sourceware.org/newlib/ this one is also complex, just like GLIBC,

bsd libc: https://github.com/freebsd/freebsd/blob/master/COPYRIGHT

**Comment by GitMensch** [comment]
Yes, drop old CPU support for OCC - if someone really needs this on Windows then MinGW-GCC does a better job for those old systems and will have the burden to keep the support for those. Removing it leads to a cleaner OCC (its complete compilation and size is one of the big advantages to compilers like GCC taking several hours to fully build and also it doesn't need to work in embedded systems like GCC does) and less time needed to be invested.

**Comment by LADSoft** [comment]
yeah I really didn't want to duplicate all the testing twice :).   

I looked at Visual C code generation and it actually mixes floating point instructions and SSE instructions in odd ways, due to the need to optimize around the fact that function return values for floating point are in st0 I suppose.   I guess I need to do that to allow use with random DLLs...

I've got the compiler rewritten and am working on debugging the changes!   should have that done in a couple of days and then start in on the SSE work...

**Comment by LADSoft** [comment]
the backend of the compiler has been rewritten and I'm working on converting the compiler itself to generate SSE instructions instead of the old FP-style instructions

**Comment by GitMensch** [comment]
Thank you for the update. Would it be useful to re-run the GnuCOBOL tests already or is it more reasonable to wait some days/weeks?

**Comment by LADSoft** [comment]
better to wait a little while.   It seems a little different now, closer to visual c's output,  but I haven't evaluated fully.   Meanwhile I still have to update the run time library...

**Comment by LADSoft** [comment]
I pushed a version which should fully address this issue.   It is much cleaner now.

Among other things I found a long standing problem with how the unary minus and complement operators work...   many of the tflt2int problems stemmed from that.

next up I'm going to fix #220.   But that can wait until tomorrow :)

**Comment by GitMensch** [comment]
For GnuCOBOL those results are fine now, just asking: Do the tests from the numeric testsuite "pass" now or only similar to VisualC's result? Is anything (if not *maybe* its should) included in the automated tests?
